org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -595,9 +597,16 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
 	@Override
 	public CompletableFuture<Object> executeCommand(ExecuteCommandParams params) {
 		debugTrace(">> workspace/executeCommand " + (params == null ? null : params.getCommand()));
-		return computeAsync((monitor) -> {
-			return commandHandler.executeCommand(params, monitor);
-		});
+		// see https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/2799
+		// Handle the paste event synchronously in the main thread
+		if (IDelegateCommandHandler.JAVA_EDIT_HANDLE_PASTE_EVENT.equals(params.getCommand())) {
```

Comment:
We should just be aware that this is probably going to grow. I could already see us wanting to do the same for smart semicolon insertion.

Reply:
Fixed